cunumeric/array.py

```diff
@@ -955,12 +955,101 @@ def byteswap(self, inplace=False):
             return self.convert_to_cunumeric_ndarray(numpy_array, stacklevel=3)
 
-    @unimplemented
-    def choose(self, choices, out, mode="raise"):
-        numpy_array = self.__array__(stacklevel=3).choose(
-            choices=choices, out=out, mode=mode
-        )
-        return self.convert_to_cunumeric_ndarray(numpy_array, stacklevel=3)
+    def choose(self, choices, out=None, mode="raise", stacklevel=2):
+        a = self.convert_to_cunumeric_ndarray(self)
+        if out is not None:
+            out = out.convert_to_cunumeric_ndarray(out)
+        # choices can be passed as array or tuple of arrays.
+        # in case of the tuple, we will pass each array separately
+        is_tuple = isinstance(choices, tuple)
+        if is_tuple:
+            n = len(choices)
+            ch_dtype = choices[0].dtype
+        else:
+            n = choices.shape[0]
+            ch_dtype = choices.dtype
+
+        if self.dtype != int:
+            raise TypeError("a array should be ineger type")
+        if mode == "raise":
+            if (a < 0).any() | (a >= n).any():
+                raise ValueError("invalid entry in choice array")
+        elif mode == "wrap":
+            a = a % n
+        elif mode == "clip":
+            a = a.clip(0, n - 1)
+        else:
+            raise ValueError(
+                "mode={mode!r} not understood. Must"
+                " be 'raise', 'wrap', or 'clip'"
+            )
+
+        # we need to broadcast all arrays in choices with
+        # input and output arrays
+        if out is not None:
+            out_shape = broadcast_shapes(a.shape, choices[0].shape, out.shape)
+        else:
+            out_shape = broadcast_shapes(a.shape, choices[0].shape)
+
+        for c in choices:
+            out_shape = broadcast_shapes(out_shape, c.shape)
+
+        # if there is not output specified, create one
+        if out is None:
+            # no output, create one
+            out = ndarray(
+                shape=out_shape,
+                dtype=ch_dtype,
+                stacklevel=stacklevel + 1,
+                inputs=(a),
+            )
+        else:
+            # if output is provided, it's shape should be the same as out_shape
+            if out.shape != out_shape:
+                raise ValueError(
+                    "non-broadcastable output operand with shape "
+                    + str(out.shape)
+                    + " doesn't match the broadcast shape "
+                    + str(out_shape)
+                )
+
+        if is_tuple:
+            ch_tuple = (choices[0].convert_to_cunumeric_ndarray(choices[0]),)
+            for i in range(1, len(choices)):
+                ch_tuple = ch_tuple + (
+                    choices[i].convert_to_cunumeric_ndarray(choices[i]),
+                )
+            ch = (ch_tuple[0]._thunk,)
+            for i in range(1, len(choices)):
+                ch = ch + (ch_tuple[i]._thunk,)
```

Comment:
A cleaner way to do this is:
```
ch = tuple(self.convert_to_cunumeric_ndarray(c)._thunk for c in choices)
```
Please prefer this method instead of building up the tuples with `+=` when possible. It's more concise and easier to understand.

Reply:
Sure, thanks!